[AERIE-1835] Start and End TimingConstraints in CoexistenceGoal

mattdailis opened the pull request:
* **Tickets addressed:** AERIE-1835
* **Review:** By file (for now)  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** squash and merge (for now)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->

This PR allows specifying the start or end timing of an activity relative to its corresponding window. Here's an example, where the new activity should start exactly five minutes after the end of the existing activity.

```Typescript
export default () => Goal.CoexistenceGoal({
  forEach: ActivityExpression.ofType(ActivityTypes.GrowBanana),
  activityTemplate: ActivityTemplates.PeelBanana({peelDirection: "fromStem"}),
  startsAt: TimingConstraint.singleton(WindowProperty.END).plus(5 * 60 * 1000 * 1000)
})
```

Here's another example, where the new activity should start anywhere between the end of the existing activity, and five minutes after. (note the `startsWithin` and `range`

```Typescript
export default () => Goal.CoexistenceGoal({
  forEach: ActivityExpression.ofType(ActivityTypes.GrowBanana),
  activityTemplate: ActivityTemplates.PeelBanana({peelDirection: "fromStem"}),
  startsWithin: TimingConstraint.range(WindowProperty.END, Operator.PLUS, 5 * 60 * 1000 * 1000)
})
```

~.☝️ I'm currently debugging why the `range` case results in the secant solver diverging when trying to place an uncontrollable duration activity... 🤔.~

That bug has been documented as AERIE-1904, and shouldn't block this PR.

TODO before merging:
- [x] Add syntax that allows users to have zero operations, if they want (either just a WindowProperty, or allow singleton to take a window property with no operation)

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
I have updated tests in `SchedulingDSLCompilationServiceTests`, and once those passed, I also updated tests in `SchedulingIntegrationTests`.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->

- [ ] TODO before merging: Document new feature in the wiki

## Future work
<!-- What next steps can we anticipate from here, if any? -->
No concrete next steps have been identified.

mattdailis committed:
Track TimeExpression operations in List instead of Map

mattdailis committed:
Define parser for TimingConstraints

mattdailis committed:
Translate TimingConstraints from AST to Goals

mattdailis committed:
Add TimingConstraints to scheduling eDSL

mattdailis committed:
Add tests for TimingConstraints

mattdailis committed:
Update singleton to allow omitting the operator

mattdailis committed:
Add null check for controllable duration parameter

It's possible for the controllable duration argument to be omitted. We
should not throw a null pointer exception in that case.

patkenneally reviewed (approved):
I ran the system with your tests and all worked as expected.

pcrosemurgy commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GoalBuilder.java:
```diff
@@ -48,11 +48,29 @@ public static Goal goalOfGoalSpecifier(
+        final var timeExpression = new TimeExpressionRelativeFixed(
+            startConstraint.windowProperty(),
+            startConstraint.singleton(),
+            "George");
```
Is `"George"` supposed to be here? At first I thought this was some test file – maybe there's context I'm missing?

pcrosemurgy replied:
I'd like to lobby for this to be `"Paul"`

mattdailis replied:
😆 Really good point, this was me being amused that there's a useless field. A lot of nodes in the scheduler have name fields, perhaps for debugging purposes, but we should probably consider eliminating them.

I've been operating off of the Weasleys, but I'd be happy to pivot to the Beatles 😝

mattdailis replied:
I've removed name from `TimeExpressionRelativeFixed` 👍

pcrosemurgy replied:
Awesome, thanks Matt :)

pcrosemurgy commented on scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/timeexpressions/TimeExpression.java:
```diff
@@ -24,13 +26,17 @@ public abstract class TimeExpression {
 
 
-  protected LinkedHashMap<TimeUtility.Operator, Duration> operations = new LinkedHashMap<>();
+  protected List<Pair<TimeUtility.Operator, Duration>> operations = new ArrayList<>();
```
This can be made `final` :)

mattdailis replied:
Done

mattdailis committed:
Remove name from TimeExpressionRelativeFixed

pranav-super commented on scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts:
```diff
@@ -69,8 +69,8 @@ export class Goal {
   }
 }
 
-type StartTimingConstraint = { startsAt: SingletonTimingConstraint } | { startsWithin: RangeTimingConstraint }
```
I recently rebased and in my version of scheduler-edsl-fluent-api.ts, I never had these lines. I'm not sure if that's a problem with this commit or on my end or if it'll cause merging issues.

mattdailis replied:
I think you might be looking at a line that I changed twice in this PR - I first introduced `SingletonTimingConstraint`, and then expanded it to have a `SingletonTimingConstraintNoOperation`.

mattdailis commented:
Planning to merge if/when the tests pass :+1:

mattdailis committed:
Adjust imports to work in monaco

Two main changes here: prefix filenames with file:/// and fix a typo
(scheduler-edsl-fluent-api was misspelled).

Stripping the .js extension is less necessary, but since we're not
doing node-style imports in user code, there is no reason to keep it.

dyst5422 commented on scheduler-server/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts:
```diff
@@ -1,5 +1,5 @@
-import * as AST from './scheduler-ast.js';
```
This change doesn't work for the server-side. You have to specify the .js extension when running in ESM on Node.

mattdailis replied:
Thanks for weighing in - I was just about to poke you about this.

We aren't using Node module resolution, right? If I understand correctly, we're using our custom resolution in the `aerie-ts-user-code-runner`, where we strip the extension.

dyst5422 replied:
This assumes that we don't need to import these files anywhere in our code that runs outside of the vm

mattdailis replied:
That's a good point - I'll try putting `.js` endings back in and double check that it still works in monaco.

mattdailis replied:
Pushed a commit that adds `.js` to all our imports in the scheduling ts files (I had added a few without `.js` in previous PRs).

mattdailis committed:
Use Node-style imports consistently in scheduling eDSL

mattdailis committed:
Add doc comments to TimingConstraint.singleton and range

mattdailis commented:
I'm done making changes to this PR. @pcrosemurgy let me know if you have any qualms about merging it. @patkenneally assuming this makes it into 0.12.0, I'll work on updating the docs in parallel with your release process

pcrosemurgy reviewed (approved):
Tests passing for me!

mattdailis commented:
Docs have been [updated](https://github.com/NASA-AMMOS/aerie/wiki/Scheduling-Guide/_compare/293888edc47331e22695ec9ee71383794bcf27ce...cec753856f0ba10f09a663493acebfc19232e2bd)
